Add Wbemcli classes needed to query WMI

dbwiddis opened the pull request.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/Wbemcli.java:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        public HRESULT Get(BSTR wszName, NativeLong lFlags, VARIANT.ByReference pVal, Pointer pvtType,
+                LongByReference plFlavor) {
```
According to the documents from microsoft this is the definition:
```
HRESULT Get(
  LPCWSTR wszName,
  long    lFlags,
  VARIANT *pVal,
  CIMTYPE *pType,
  long    *plFlavor
);
```

So I would use String fro `wszName` and wrap it into a `WString` for the call. Microsoft defines the sizes of the base types in:

https://msdn.microsoft.com/de-de/library/cc953fe1.aspx

So `lFlags` can be bound as int and `plFlavor` is a `IntByReference`. The `LongByReference` is to big. `pType` (it is not pvtType) is defined as:

```
typedef long CIMTYPE;
```

So it is not a pointer!

dbwiddis replied:
Wouldn't `CIMTYPE *pType` mean that `pType` is a pointer to `CIMTYPE` so a Pointer is valid here (`IntByReference` would be better as you note for `long *plFlavor`).  I have read the value from this call using `Pointer.getInt()` and it has been the CIMTYPE.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/Wbemcli.java:
```diff
@@ -0,0 +1,172 @@
+            super(pvInstance);
+        }
+
+        public HRESULT Next(NativeLong lTimeOut, NativeLong uCount, PointerByReference ppObjects,
```
Same comment as in `IWbemClassObject` applies - is it expected, that this method is called on a non windows platform? if not, `NativeLong` can be substituted with `int`.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java:
```diff
@@ -0,0 +1,783 @@
+        // Connect to WMI through the IWbemLocator::ConnectServer method
+        // Connect to the namespace with the current user and obtain pointer
+        // pSvc to make IWbemServices calls.
+        HRESULT hres = loc.ConnectServer(new BSTR(namespace), null, null, null, null, null, null, pSvc);
```
Don't use the `BSTR` constructor. `BSTR` need to be allocated via `OleAuto.INSTANCE.SysAllocString` and they need to be freed with `OleAuto.INSTANCE.SysFreeString`.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/Wbemcli.java:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        public HRESULT ConnectServer(BSTR strNetworkResource, BSTR strUser, BSTR strPassword, BSTR strLocale,
+                NativeLong lSecurityFlags, BSTR strAuthority, Pointer pCtx, PointerByReference ppNamespace) {
```
You could use String instead of BSTR and allocate + release the BSTRs inside the function (see below for allocation of BSTR).

For `pCtx` should this really be a plain Pointer?

dbwiddis replied:
For `pCtx` it would probably be an [IWbemContext](https://docs.microsoft.com/en-us/windows/desktop/api/WbemCli/nn-wbemcli-iwbemcontext) object.  But in all the sample code I have seen on MS sites for WMI connections (both local and remote) it's been `null`.   Reading the docs, it looks like `IUnknown` (or is there a ByReference version of it) may be correct?

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java:
```diff
@@ -0,0 +1,783 @@
+    /**
+     * Initializes COM library and sets security to impersonate the local user
+     */
+    public static void initCOM() {
```
This looks dangerous if you combine it with other COM calling code.

dbwiddis replied:
I agree it looks dangerous but I'm not sure how else to handle it.  I can ignore "already initialized" errors but then if I uninit later, it will kill the external program's COM code.  If I don't uninit, it's a resource leak.

matthiasblaesing replied:
Make the user responsible for initializing COM. Else you need to guess how the user will use it. The COM Utilities have offer both options. COMThread handles the creation of a thread for dispatching COM Threads with correctly initialized COM. `Factory` uses that, while `ObjectFactory` allows creation of call proxies without a different thread.

I found it better to handle COM initialization myself, so I prefer to use the `ObjectFactory`.

There is a problem with your code - when calls to the utility methods are done from different threads, COM will only be initialized once. This violates the requirement, that `CoInitializeEx` must be called on each thread making COM calls.

dbwiddis replied:
I've moved the COM outside the Util class.

dbwiddis commented on contrib/platform/src/com/sun/jna/platform/win32/Ole32.java:
```diff
@@ -134,6 +137,215 @@
+    HRESULT CoSetProxyBlanket(Pointer pProxy, //
+            int dwAuthnSvc, //
+            int dwAuthzSvc, //
+            BSTR pServerPrincName, // OLECHAR
```
MS is evil!  I saw `typedef OLECHAR * BSTR` and thought I was good.  I only ever call that function with `null` so this never reared its ugly head for me... will fix.

dbwiddis commented on contrib/platform/src/com/sun/jna/platform/win32/Ole32.java:
```diff
@@ -134,6 +137,215 @@
+     *
+     *         E_OUT_OF_MEMORY Out of memory.
+     */
+    HRESULT CoInitializeSecurity(Pointer pSecDesc, NativeLong cAuthSvc, Pointer asAuthSvc, Pointer pReserved1,
```
Yep. Another case where I always used `null` and was lazy.

dbwiddis commented on contrib/platform/src/com/sun/jna/platform/win32/Ole32.java:
```diff
@@ -134,6 +137,215 @@
+     *
+     *         E_INVALIDARG One or more arguments is invalid.
+     */
+    HRESULT CoSetProxyBlanket(Pointer pProxy, //
```
It's the pointed-to value of the last argument of [IwbemLocator::ConnectServer](https://docs.microsoft.com/en-us/windows/desktop/api/wbemcli/nf-wbemcli-iwbemlocator-connectserver) which is a pointer to an [IWbemServices](https://docs.microsoft.com/en-us/windows/desktop/api/wbemcli/nn-wbemcli-iwbemservices) object.  In the code I use this `Pointer` in the `IWbemServices` initializer.  Not sure what other mapping it should be.

dbwiddis commented:
I think I've addressed all the comments except the type mapping for the last argument of `ConnectServer`.  It works as I have it with a `Pointer`.  I tried various combinations of `Unknown` and `ByReference` but ended up unable to access the multiple layers of pointers that `PointerByReference` allows.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java:
```diff
@@ -0,0 +1,700 @@
+        public Integer getInteger(T property, int index) {
+            Object o = this.propertyMap.get(property).get(index);
+            if (o == null) {
+                return 0;
```
I'm a bit irritated here - why return a primitive 0 instead of the java `null`? I would understand if you return int, but with a nullable type?

dbwiddis replied:
That's a very good question, and I went back and forth on this implementation several times.  My reasoning:
- Requiring the user to null-check every return takes away some of the value of a utility class that provides sensible defaults (zero or empty string).  The user still has the ability to test null (and even determine types if not-null) directly using `getValue()`. 
- The return type should be tied directly to the CIM type.  `getInteger()` should only be used for `uint16`, `sint32`, and `uint32` CIM types, all of which suggest a numeric value.  The only time we run into an issue is if WMI has no value for that column, and we don't check the type.
- I've also considered having different return types based on the CIM Type rather than the value type encoded in the Variant.  For example, both `String` and `uint64` return as a Variant BSTR, but the latter requires further parsing (to either `long` or possibly a `BigInteger` or `WinNT.LARGE_INTEGER`.  All of which returns to the question of how to treat parse exceptions / nulls.   I eventually settled on what you see as a happy medium.
- If I had my druthers I'd probably use Java 8's `Optional<Integer>` here.  But that's not an option (pun intended).
- All these numeric values began their life as primitives, from `VARIANT`'s various `.*value()` methods. They only got boxed when being added to the Lists.  When I wrote this, I unboxed and returned primitives, and that's why I initially wrote the defaults.
- In many cases in my application I would immediately re-box them to put into another list or map.  So since the value was already boxed in the list, I felt returning it as it was (with a simple typecast) was the most direct option and changed it.  So the code you see is a relic of the original return type but I still thought the defaults were more useful than null (since the user still has access to test null).
 
From the perspective of a user who cares about null, they could write:

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java:
```diff
@@ -0,0 +1,700 @@
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                OleAuto.INSTANCE.SysFreeString(WQL);
```
I would not preallocated `WQL` and do the allocation in the query. At least I would measure if it really is a significant part of the call cost. This construction looks fragile.

dbwiddis replied:
The allocation and release take an average of 33 microseconds for both calls.  Assuming (generously) I could kill one of them that'd save 16 microseconds per invocation.  So, yeah, not too significant. :)

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java:
```diff
@@ -0,0 +1,700 @@
+
+    // Track initialization of COM and Security
+    private static boolean comInitialized = false;
+    private static boolean securityInitialized = false;
```
These variables are not used anymore.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/Wbemcli.java:
```diff
@@ -0,0 +1,172 @@
+import com.sun.jna.ptr.PointerByReference;
+
+/**
+ * This header is used by Remote Desktop Services.
```
?! Not sure what the header wants to tell me?

dbwiddis replied:
Well, I stole it [from here](https://docs.microsoft.com/en-us/windows/desktop/api/wbemcli/).  I guess I can make it more useful.

dbwiddis commented on contrib/platform/src/com/sun/jna/platform/win32/Wbemcli.java:
```diff
@@ -0,0 +1,172 @@
+            HRESULT hres = Ole32.INSTANCE.CoCreateInstance(CLSID_WbemLocator, null, WTypes.CLSCTX_INPROC_SERVER,
+                    IID_IWbemLocator, pbr);
+            if (COMUtils.FAILED(hres)) {
+                Ole32.INSTANCE.CoUninitialize();
```
Yeah, I had already removed that, somehow I committed it to a different branch.  Same with the removed variables.

dbwiddis commented on contrib/platform/src/com/sun/jna/platform/win32/Wbemcli.java:
```diff
@@ -0,0 +1,172 @@
+     * Exception encountered in this class
+     */
+    @SuppressWarnings("serial")
+    class WbemcliException extends RuntimeException {
```
Actually it's used if `CoCreateInstance()` fails.

matthiasblaesing replied:
I see - thank you for the pointer, I really overlooked that case.

matthiasblaesing replied:
`COMBindingBaseObject` uses `COMUtils#checkRC` which raises a `COMException` on error - that would align `Wbemcli` with other COM bindings.

dbwiddis replied:
Good timing, I was actually just about to update this.  Based on the above, I'll
- return null from this method if it fails
- handle null in the util class with a `COMException` along with other uses of this exception.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/Wbemcli.java:
```diff
@@ -0,0 +1,172 @@
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+package com.sun.jna.platform.win32;
```
These are COM bindings could you please move this and the Util class to the `com.sun.jna.platform.win32.COM` package?

dbwiddis commented on contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java:
```diff
@@ -0,0 +1,682 @@
+        OleAuto.INSTANCE.SysFreeString(queryStr);
+        OleAuto.INSTANCE.SysFreeString(wql);
+        if (COMUtils.FAILED(hres)) {
+            svc.Release();
```
It was released there in the [sample code that I ported](https://docs.microsoft.com/en-us/windows/desktop/WmiSdk/example--getting-wmi-data-from-the-local-computer), although since I've moved out other resource releasing, may as well do that here, as well.

matthiasblaesing commented:
I would be willing to merge - shall I?

matthiasblaesing commented:
Hi Daniel, I had another look at this (sorry for the delay). Looking at the Wbem classes, they look a lot like JDBC and could be wrapped similarly, I did not pursue that further, as that might be overkill. However, I added a few changes, that you might want to consider:

https://github.com/matthiasblaesing/jna/commits/dbwiddis-wmi

Please have a look at the last three commits of the branch referenced above (If all went well, you should be able to pull directly from it). The first commit adds a bit binding sugar, the second moves WmiQuery more into a object oriented direction and the third fixes the unittests.

matthiasblaesing commented:
Thank you for your work - its appreciated. I added a slight update:

https://github.com/matthiasblaesing/jna/commit/68c8a96f5a912f3160073a6dbeb4ac437b53c6a1

I squashed the work into one feature commit and merged that to master.
